### PR TITLE
Release v1.4.0

### PR DESCRIPTION
**Pull request by bryanculver**
Request to merge `develop` into `main` for v1.4.0

# TODO

## Required
- [x] Update `pyproject.toml` and date in release notes
- [x] #2200 
- [x] #2196 
- [x] #2190 
- [x] #2130 
- [x] #1902
- [x] Review feedback from #2046 

## Ideal
- [x] #2029 
- [ ] #2202 
- [x] #2182 
- [x] #2176 
- [ ] #2172

### CODE REVIEW

File: nautobot/docs/development/style-guide.md [hunk 1]
```diff
@@ -57,9 +57,19 @@ New dependencies can be added to the project via the `poetry add` command. This
...
 
-* The combination of `nautobot.utilities.forms.BootstrapMixin`, `nautobot.extras.forms.CustomFieldModelForm` and `nautobot.extras.forms.RelationshipModelForm` is such a common use case throughout the code base that they have a helper class which combines all three at `nautobot.extras.forms.NautobotModelForm`. Use this helper class if you need the functionality from these three classes.
+* The combination of `nautobot.utilities.forms.BootstrapMixin`, `nautobot.extras.forms.CustomFieldModelFormMixin`, `nautobot.extras.forms.RelationshipModelFormMixin` and `nautobot.extras.forms.NoteModelFormMixin` is such a common use case throughout the code base that they have a helper class which combines all three at `nautobot.extras.forms.NautobotModelForm`. Use this helper class if you need the functionality from these three classes.
```
Comment (glennmatthews) [inline nautobot/docs/development/style-guide.md:62]:
```suggestion
* The combination of `nautobot.utilities.forms.BootstrapMixin`, `nautobot.extras.forms.CustomFieldModelFormMixin`, `nautobot.extras.forms.RelationshipModelFormMixin` and `nautobot.extras.forms.NoteModelFormMixin` is such a common use case throughout the code base that they have a helper class which combines all of these at `nautobot.extras.forms.NautobotModelForm`. Use this helper class if you need the functionality from these classes.
```

File: nautobot/docs/development/style-guide.md [hunk 2]
```diff
@@ -57,9 +57,19 @@ New dependencies can be added to the project via the `poetry add` command. This
...
+
+* Similarly, for filter forms, `nautobot.extras.forms.NautobotFilterForm` combines `nautobot.utilities.forms.BootstrapMixin`, `nautobot.extras.forms.CustomFieldModelFilterFormMixin`, and `nautobot.extras.forms.RelationshipModelFilterFormMixin`, and should be used where appropriate.
```
Comment (glennmatthews) [inline nautobot/docs/development/style-guide.md:64]:
```suggestion
* _Added in version 1.4_: Similarly, for filter forms, `nautobot.extras.forms.NautobotFilterForm` combines `nautobot.utilities.forms.BootstrapMixin`, `nautobot.extras.forms.CustomFieldModelFilterFormMixin`, and `nautobot.extras.forms.RelationshipModelFilterFormMixin`, and should be used where appropriate.
```

File: nautobot/docs/development/style-guide.md [hunk 3]
```diff
@@ -57,9 +57,19 @@ New dependencies can be added to the project via the `poetry add` command. This
...
+
+* Similarly, for bulk-edit forms, `nautobot.extras.forms.NautobotBulkEditForm` combines `nautobot.utilities.forms.BulkEditForm` and `nautobot.utilities.forms.BootstrapMixin` with `nautobot.extras.forms.CustomFieldModelBulkEditFormMixin`, `nautobot.extras.forms.RelationshipModelBulkEditFormMixin` and `nautobot.extras.forms.NoteModelBulkEditFormMixin`, and should be used where appropriate.
```
Comment (glennmatthews) [inline nautobot/docs/development/style-guide.md:66]:
```suggestion
* _Added in version 1.4_: Similarly, for bulk-edit forms, `nautobot.extras.forms.NautobotBulkEditForm` combines `nautobot.utilities.forms.BulkEditForm` and `nautobot.utilities.forms.BootstrapMixin` with `nautobot.extras.forms.CustomFieldModelBulkEditFormMixin`, `nautobot.extras.forms.RelationshipModelBulkEditFormMixin` and `nautobot.extras.forms.NoteModelBulkEditFormMixin`, and should be used where appropriate.
```

File: nautobot/docs/models/extras/jobhook.md [hunk 4]
```diff
@@ -0,0 +1,62 @@
+# Job Hooks
+
```
Comment (glennmatthews) [inline nautobot/docs/models/extras/jobhook.md:2]:
```suggestion

_Added in version 1.4.0_ <!-- markdownlint-disable-line MD036 -->

```

File: nautobot/docs/models/dcim/location.md [hunk 5]
```diff
@@ -0,0 +1,14 @@
+# Locations
+
+<!-- markdownlint-disable MD036 -->
+_Added in version 1.4.0_
+<!-- markdownlint-enable MD036 -->
```
Comment (glennmatthews) [inline nautobot/docs/models/dcim/location.md:5]:
```suggestion
_Added in version 1.4.0_ <!-- markdownlint-disable-line MD036 -->
```

File: nautobot/docs/models/dcim/locationtype.md [hunk 6]
```diff
@@ -0,0 +1,27 @@
+# Location Types
+
+<!-- markdownlint-disable MD036 -->
+_Added in version 1.4.0_
+<!-- markdownlint-enable MD036 -->
```
Comment (glennmatthews) [inline nautobot/docs/models/dcim/locationtype.md:5]:
```suggestion
_Added in version 1.4.0_ <!-- markdownlint-disable-line MD036 -->
```

File: nautobot/docs/models/extras/note.md [hunk 7]
```diff
@@ -0,0 +1,5 @@
+# Note
+
```
Comment (glennmatthews) [inline nautobot/docs/models/extras/note.md:2]:
```suggestion

_Added in version 1.4.0_ <!-- markdownlint-disable-line MD036 -->

```

File: nautobot/docs/plugins/development.md [hunk 8]
```diff
@@ -259,9 +263,90 @@ class SiteAnimalCount(PluginTemplateExtension):
...
+                "title": "Plugin Tab 2",
+                "url": reverse("plugins:example_plugin:device_detail_tab_2", kwargs={"pk": self.context["object"].pk}),
+            },
+        ]
+
+template_extensions = [DeviceExtraTabs, SiteAnimalCount]
+```
+
+#### Adding Extra Tabs
+
```
Comment (glennmatthews) [inline nautobot/docs/plugins/development.md:300]:
```suggestion

_Added in version 1.4.0_ <!-- markdownlint-disable-line MD036 -->

```

File: nautobot/docs/plugins/development.md [hunk 9]
```diff
@@ -1009,3 +1095,54 @@ plugin_name/                   # "nautobot_animal_sounds"
         - models/
           - object_model.html  # "animal.html"
 ```
+
+## Overriding Existing Functionality
+
+### Replacing Views
+
```
Comment (glennmatthews) [inline nautobot/docs/plugins/development.md:1102]:
```suggestion

_Added in version 1.4.0_ <!-- markdownlint-disable-line MD036 -->

```

File: nautobot/docs/plugins/development.md [hunk 10]
```diff
@@ -1009,3 +1095,54 @@ plugin_name/                   # "nautobot_animal_sounds"
...
+
+override_views = {
+    "dcim:device": DeviceViewOverride.as_view(),
+}
+```
+
+## Note URL Endpoint
+
+New in Nautobot 1.4 is support for models that inherit from `PrimaryModel` and `OrganizationalModel` to have notes associated. In order to utilize this new feature you will need to add the endpoint to `urls.py`. Here is an option to be able to support 1.4 and older versions of Nautobot:
```
Comment (glennmatthews) [inline nautobot/docs/plugins/development.md:1128]:
```suggestion

_Added in version 1.4.0_ <!-- markdownlint-disable-line MD036 -->

Models that inherit from `PrimaryModel` and `OrganizationalModel` can have notes associated. In order to utilize this new feature you will need to add the endpoint to `urls.py`. Here is an option to be able to support both 1.4+ and older versions of Nautobot:
```

File: nautobot/docs/plugins/index.md [hunk 11]
```diff
@@ -8,10 +8,10 @@ The Nautobot plugin architecture allows for plugins to do any or all of the foll
 
 ### Extend the existing Nautobot UI
 
-* **Add content to existing model detail views.** A plugin can inject custom HTML content within the view of a core Nautobot model. This content can appear in the left column, right column, or full width of the page, and can also include custom buttons at the top of the page.
 * **Add navigation menu items.** A plugin can extend the navigation menus with new links and buttons or even entirely new menus.
 * **Add a banner.** A plugin can add a custom banner to the top of any appropriate views.
 * **Add home page content.** A plugin can add custom items or custom panels to the Nautobot home page.
+* **Add content to existing model detail views.** A plugin can inject custom HTML content within the view of a core Nautobot model. This content can appear in the left column, right column, or full width of the page, include custom buttons at the top of the page and include extra tabs which appear at the end of the object detail tabs list.
```
Comment (glennmatthews) [inline nautobot/docs/plugins/index.md:14]:
```suggestion
* **Add content to existing model detail views.** A plugin can inject custom HTML content within the view of a core Nautobot model. This content can appear in the left column, right column, or full width of the page, include custom buttons at the top of the page and (in Nautobot 1.4 and later) include extra tabs which appear at the end of the object detail tabs list.
```

File: nautobot/docs/plugins/index.md [hunk 12]
```diff
@@ -21,6 +21,7 @@ The Nautobot plugin architecture allows for plugins to do any or all of the foll
 * **Register additional Jinja2 filters.** A plugin can define custom Jinja2 filters to be used in computed fields, webhooks, custom links, and export templates.
 * **Populate extensibility features in the database.** A plugin can add content to the Nautobot database when installed, such as automatically creating new custom fields, relationships, and so forth.
 * **Add additional secrets providers.** A plugin can add support for retrieving [Secret](../models/extras/secret.md) values from additional sources or external systems.
+* **Override already-defined views.** A plugin can define a view which can be set to override a view from the core set of views or another plugin's view.
```
Comment (glennmatthews) [inline nautobot/docs/plugins/index.md:24]:
```suggestion
* **Override already-defined views.** _Added in version 1.4.0_ A plugin can define a view which can be set to override a view from the core set of views or another plugin's view.
```

File: nautobot/docs/rest-api/filtering.md [hunk 13]
```diff
@@ -48,11 +52,16 @@ To filter results by a custom field value, prepend `cf_` to the custom field nam
...
+
 Custom fields can be mixed with built-in fields to further narrow results. When creating a custom string field, the type of filtering selected (loose versus exact) determines whether partial or full matching is used.
 
+Custom fields can use the lookup expressions listed in the next section by prepending `cf_` to the custom field `name` (and not the `slug`) followed by the required lookup type (see below).
+
 ## Lookup Expressions
 
-Certain model fields also support filtering using additional lookup expressions. This allows
+Certain model fields (including custom fields of type `text`, `url`, `select`, `integer`, and `date`) also support filtering using additional lookup expressions. This allows
```
Comment (glennmatthews) [inline nautobot/docs/rest-api/filtering.md:64]:
```suggestion
Certain model fields (including, in Nautobot 1.4 and later, custom fields of type `text`, `url`, `select`, `integer`, and `date`) also support filtering using additional lookup expressions. This allows
```

File: nautobot/docs/rest-api/filtering.md [hunk 14]
```diff
@@ -48,11 +52,16 @@ To filter results by a custom field value, prepend `cf_` to the custom field nam
 GET /api/dcim/sites/?cf_foo=123
 ```
 
+!!! note
+    For custom field filters, due to historical details of implementation, only a single filter value can be specified when matching a given field. In other words, in the above example, you could _not_ add `&cf_foo=456` to the query in order to get all sites where custom field `foo` is 123 _or_ 456; instead you would need to run two separate queries. This restriction does not apply to custom field filters using lookup expressions (next section) and will likely be changed in a future major version of Nautobot.
+
 Custom fields can be mixed with built-in fields to further narrow results. When creating a custom string field, the type of filtering selected (loose versus exact) determines whether partial or full matching is used.
 
+Custom fields can use the lookup expressions listed in the next section by prepending `cf_` to the custom field `name` (and not the `slug`) followed by the required lookup type (see below).
```
Comment (glennmatthews) [inline nautobot/docs/rest-api/filtering.md:60]:
```suggestion
_Added in version 1.4.0_: Custom fields can use the lookup expressions listed in the next section by prepending `cf_` to the custom field `name` (and not the `slug`) followed by the required lookup type (see below).
```

**Review by glennmatthews — COMMENTED** [review]
Some docs cleanup suggestions.

File: nautobot/docs/additional-features/template-filters.md [hunk 15]
```diff
@@ -117,6 +117,18 @@ Humanize speeds given in Kbps.
 {{ speed_value | humanize_speed }}
 ```
 
+### hyperlinked_object
+
```
Comment (glennmatthews) [inline nautobot/docs/additional-features/template-filters.md:121]:
```suggestion

_Added in version 1.4.0_ <!-- markdownlint-disable-line MD036 -->

```

File: nautobot/docs/configuration/optional-settings.md [hunk 16]
```diff
@@ -943,6 +920,25 @@ If [`STORAGE_BACKEND`](#storage_backend) is not defined, this setting will be ig
 
 ---
 
+## STRICT_FILTERING
+
+<!-- markdownlint-disable MD036 -->
+_Added in version 1.4.0_
+<!-- markdownlint-enable MD036 -->
```
Comment (glennmatthews) [inline nautobot/docs/configuration/optional-settings.md:927]:
```suggestion
_Added in version 1.4.0_ <!-- markdownlint-disable-line MD036 -->
```

File: nautobot/docs/configuration/optional-settings.md [hunk 17]
```diff
@@ -975,3 +971,54 @@ Environment Variables:
...
 * `NAUTOBOT_DATETIME_FORMAT`
 * `NAUTOBOT_SHORT_DATETIME_FORMAT`
+
+---
+
+## UI_RACK_VIEW_TRUNCATE_FUNCTION
+
+<!-- markdownlint-disable MD036 -->
+_Added in version 1.4.0_
+<!-- markdownlint-enable MD036 -->
```
Comment (glennmatthews) [inline nautobot/docs/configuration/optional-settings.md:981]:
```suggestion
_Added in version 1.4.0_ <!-- markdownlint-disable-line MD036 -->
```

**Review by glennmatthews — APPROVED** [review]
:shipit: :shipit: :shipit: :shipit:

**Review by jathanism — APPROVED** [review]

**Pull request merged**